sdks/java/io/hive/src/main/java/org/apache/beam/sdk/io/hive/HiveIO.java

```diff
@@ -0,0 +1,474 @@
+
+    abstract Builder toBuilder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setMetastoreUri(String metastoreUri);
+      abstract Builder setDatabase(String database);
+      abstract Builder setTable(String table);
+      abstract Builder setFilter(String filter);
+      abstract Builder setCoder(Coder<HCatRecord> coder);
+      abstract Builder setSplitId(Integer splitId);
+      abstract Builder setContext(ReaderContext context);
+      abstract Read build();
+    }
+
+    public Read withMetastoreUri(String metastoreUri) {
+      return toBuilder().setMetastoreUri(metastoreUri).build();
+    }
+
+    public Read withDatabase(String database) {
+      return toBuilder().setDatabase(database).build();
+    }
+
+    public Read withTable(String table) {
+      return toBuilder().setTable(table).build();
+    }
+
+    public Read withFilter(String filter) {
+      return toBuilder().setFilter(filter).build();
+    }
+
+    protected Read withSplitId(Integer splitId) {
+      checkArgument(splitId >= 0);
+      return toBuilder().setSplitId(splitId).build();
+    }
+
+    protected Read withContext(ReaderContext context) {
+      return toBuilder().setContext(context).build();
+    }
+
+    @Override
+    public PCollection<HCatRecord> expand(PBegin input) {
+      return input.apply(org.apache.beam.sdk.io.Read.from(new BoundedHiveSource(this)));
+    }
+
+    @Override
+    public void validate(PipelineOptions options) {
+      checkNotNull(getTable(), "table");
+      checkNotNull(getMetastoreUri(), "metastoreUri");
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      super.populateDisplayData(builder);
+      builder.add(DisplayData.item("metastoreUri", getMetastoreUri()));
+      builder.add(DisplayData.item("table", getTable()));
+      builder.addIfNotNull(DisplayData.item("database", getDatabase()));
+      builder.addIfNotNull(DisplayData.item("filter", getFilter()));
+    }
+  }
+
+  /**
+   * A Hive {@link BoundedSource} reading {@link HCatRecord} from  a given instance.
+   */
+  @VisibleForTesting
+  static class BoundedHiveSource extends BoundedSource<HCatRecord> {
+    private Read spec;
+
+    private BoundedHiveSource(Read spec) {
+      this.spec = spec;
+    }
+
+    @Override
+    public Coder<HCatRecord> getDefaultOutputCoder() {
+      return HCatRecordSimpleCoder.of(spec.getMetastoreUri());
+    }
+
+    @Override
+    public void validate() {
+      spec.validate(null);
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      spec.populateDisplayData(builder);
+    }
+
+    @Override
+    public BoundedReader<HCatRecord> createReader(PipelineOptions options) {
+      return new BoundedHiveReader(this);
+    }
+
+    @Override
+    public long getEstimatedSizeBytes(PipelineOptions pipelineOptions) throws
+    IOException, NoSuchObjectException, TException {
```

Comment:
Why so many throws declarations? I think `getEstimatedSizeBytes` is only allowed to throw `IOException`, right?